eclipse-ee4j/jaxrs-api
# TCK Migration: Move first set of REST tck tests from jakartaee-tck

The changes are based on https://github.com/hantsy/jakartaee9-starter-boilerplate/blob/master/docs/arq-glassfish.md to using arquillian(glassfish as container)  & Junit5 to run  tests. 
To test the changes :

export JAVA_HOME=<JDK11_HOME>
export M2_HOME=<MAVEN_HOME>
export PATH=$M2_HOME/bin:$JAVA_HOME/bin:$PATH

jaxrs-tck/
mvn clean install

jersey-tck/
mvn clean verify -Parq-glassfish-managed

The tests migrated :
jakartaee-tck/src/com/sun/ts/tests/jaxrs/api/rs/core/responsebuilder (15)
jakartaee-tck/src/com/sun/ts/tests/jaxrs/api/rs/core/responseclient (85)
jakartaee-tck/src/com/sun/ts/tests/jaxrs/api/rs/core/uribuilder (126)
jakartaee-tck/src/com/sun/ts/tests/jaxrs/ee/rs/core/request (28)
jakartaee-tck/src/com/sun/ts/tests/jaxrs/ee/rs/client/asyncinvoker (147)

**Current status** : 11 test failures in ee/rs/client/asyncinvoker tracked at https://github.com/eclipse-ee4j/jaxrs-api/issues/1020
JAXRSClientIT.traceTest:
JAXRSClientIT.traceWhileServerWaitTest
JAXRSClientIT.traceWithCallbackThrowsNoWebApplicationExceptionForResponseTest:
JAXRSClientIT.traceWithCallbackWhileServerWaitTest:
JAXRSClientIT.traceWithClassThrowsNoWebApplicationExceptionForResponseTest:
JAXRSClientIT.traceWithGenericTypeResponseWhileServerWaitTest:
JAXRSClientIT.traceWithGenericTypeThrowsNoWebApplicationExceptionForResponseTest:
JAXRSClientIT.traceWithResponseClassWhileServerWaitTest:
JAXRSClientIT.traceWithGenericTypeStringWhileServerWaitTest:
JAXRSClientIT.traceWithStringCallbackWhileServerWaitTest:
JAXRSClientIT.traceWithStringClassWhileServerWaitTest:

**Resolved Issues**:

-   set properties in arquillian.xml/pom.xml : system properties are now set in jersey-tck/pom.xml
-   Verify using the glassfish container in arquillian is the correct way to test the standalone TCK.
-   Currently the Jersey jars in the pom are used on the client, but the server side jars are used from the GF. The jersey jars should be replaced in the glassfish bundle with the ones that need to be tested : After the glassfish bundle is unpacked the jersey jars are explicitly downloaded and replaced at glassfish/modules so the expected jersey jars are tested.
-   JAX-RS 3.1 won't support JDK 8, so need to use GF 6.1 and run it on JDK 11.  : Using JDK11 & glassfish 6.1 currently.
-   Correctly Use TestUtil.java or equivalent for logging based on traceflag parameters : Improved logging , using new system variable junit.log.traceflag for trace logs, printing the logs as in the original tck run.
- use web.xml for deployment, currently it is using hardcoded web.xml string : The web.xml.template files are moved to resource directory src/main/resources. The same is obtained as string, replaced "servlet_adaptor" with System variable and added as string web xml for deployment. 

 **Pending items** :
- Create common deployment as mentioned by Ondre in slack channel https://eclipsefoundationhq.slack.com/archives/C0131MLD538/p1627033972014200?thread_ts=1624542758.005000&cid=C0131MLD538 
- code common across TCK, can be pulled as jars, use these jars across separated TCKs with a maven dependency (Quote from Gurus review https://github.com/eclipse-ee4j/jaxrs-api/pull/1002#discussion_r684979427)
